OlinkAnalyze/R/olink_lod.R

```diff
@@ -0,0 +1,233 @@
+#' Calculate LOD using Negative Controls or Fixed LOD
+#'
+#' @param data npx data file
+#' @param lod_file_path location of lod file from Olink (only needed if lod_method = "FixedLOD")
+#' @param lod_method method for calculating LOD using either "FixedLOD" or negative controls ("NCLOD")
+#'
+#' @return A dataframe with 2 additional columns, LOD and PCNormalizedLOD. When Normalization = "Plate Control", LOD and PCNormalizedLOD are equivalent
+#' @export
+#'
+olink_lod <- function(data, lod_file_path = NULL, lod_method = "NCLOD"){
+
+  # store original column names of `data` to restore them later
+  original_columns <- names(data)
+
+  # check the type of LOD calculation to perform and compute or extract:
+  # LodNPX, LODCount and LodMethod
+
+  lod_methods <- list(fix_lod = "FixedLOD",
+                      nc_lod = "NCLOD")
+
+  if (lod_method == lod_methods$fix_lod) {
+
+    if(missing(lod_file_path)) {
+      stop(paste0("lod_file_path must be specified for lod_method = \"",
+                  lod_methods$fix_lod, "\""))
+    }
+
+    lod_file <- read.table(file = lod_file_path, sep = ";", header = TRUE)
+    lod_data <- olink_fixed_lod(data_analysis_ref_id = data$DataAnalysisRefID,
+                                lod_file = lod_file)
+
+  } else if (lod_method == lod_methods$nc_lod) {
+
+    lod_data <- olink_nc_lod(data = data)
+
+  } else {
+
+    stop(paste0("lod_method must be one of \"", lod_methods$fix_lod, "\" or \"",
+                lod_methods$nc_lod, "\""))
+
+  }
+
+  # If NPX is intensity normalized, than intensity normalize LOD
+  data <- int_norm_count(
```

Comment:
Intensity normalized NPX is compute as NPX=ExtNPX - Plate_Median_NPX
PC normalized NPX is computed as PCnormalizedNPX = ExtNPX - PC_median